Adds a chart for the external-dns service

rk295 opened the pull request:
Hi,

This is my first chart, so please go easy. I've tried to cover all the options provided by `external-dns` with the exception of the `file` ones because I couldn't figure out how to make that work nicely (suggestions welcome). 

I only have AWS & R53 in which to test, but it is working fine for me, will obviously require more testing though.

rk295 committed:
initial commit

sstarcher commented:
@rk295 I believe this chart meets the requirements to be in the stable folder.
https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements

sstarcher commented:
Could you add support for podAnnotations.  This would enable people to use kube2iam to supply the permissions to the pod.

https://github.com/kubernetes/charts/blob/bcb4436d85a30f9912fd59dde502407fe2e36f23/stable/spartakus/templates/deployment.yaml#L14

* Deployment
```
spec:
  replicas: {{ .Values.replicaCount }}
  template:
    metadata:
    {{- if .Values.podAnnotations }}
      annotations:
{{ toYaml .Values.podAnnotations | indent 8}}
```

* Values
```
## Annotations to be added to pods
##
podAnnotations: {}
```

sstarcher commented:
After taking a closer look it should this will also need a ServiceAccount, ClusterRoleBindings as it is using the kubernetes api and RBAC is enabled in 1.6 by default.

I don't know what specific permissions this will need though.
https://github.com/kubernetes/charts/blob/2f5852105ed27f1c2ffb701b2482a83834697d7b/incubator/istio/templates/rbac.yaml#L3

sstarcher commented:
As for the file, you can support the azureConfigFile via a ConfigMap+Volume Mount

Something like the following

* ConfigMap
```
{{ if .Values.azureConfig -}}
apiVersion: v1
kind: ConfigMap
data: |
{{ toJson .Values.azureConfig | indent 2 }}
{{- end }}
```
https://github.com/kubernetes/charts/blob/224b2f7a873173b51212b8f36cd4130d3b789adf/stable/grafana/templates/configmap.yaml#L13

* Values
```
# azureConfig: ""
```

* Deployment
```
    spec:
      containers:
        - name: blah
{{ if .Values.azureConfig -}}
          volumeMounts:
            - name: config-volume
              mountPath: "/etc/azure"
      volumes:
        - name: config-volume
          configMap:
            name: {{ template "fullname" . }}-config
{{- end }}
```

linki commented on incubator/external-dns/Chart.yaml:
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
+description: Deploys external-dns into a kubernetes cluster
+name: external-dns
+version: 0.0.1
```
We can enrich the yaml similar to https://github.com/kubernetes/charts/blob/master/stable/mediawiki/Chart.yaml. This allows us to have both the chart version and external-dns's version in here.

rk295 commented:
@linki & @sstarcher sure thing, I think I can address most of the changes above. The one I'm a little out of my depth about is the RBAC one, I have no knowledge of RBAC in K8s and have no cluster with it enabled to test. So I think for now I'll be able to address the other comments but not that one. 

Up to you how to play it from here, shall I go ahead and fix the others or shall I close the PR for now until somebody can write the RBAC related templates?

linki commented on incubator/external-dns/values.yaml:
```diff
@@ -0,0 +1,68 @@
+source:
+  fake: false
+  ingress: true
+  service: true
```
This could maybe be done via
```yaml
  sources:
  - service
  - ingress
```
and then just iterate over it writing out `--source=${i}` lines in the args section of the deployment template?

linki commented on incubator/external-dns/templates/deployment.yaml:
```diff
@@ -0,0 +1,70 @@
+    spec:
+      containers:
+      - name: {{ .Chart.Name }}
+        image: "{{- if .Values.image.repository }}{{ .Values.image.repository }}/{{- end -}}{{.Values.image.name}}:{{ .Values.image.tag }}"
```
I think `.Values.image.repository` doesn't need to be guarded. If someone has it on dockerhub she could just use `repository: docker.io`?

linki commented on incubator/external-dns/values.yaml:
```diff
@@ -0,0 +1,68 @@
+# azureResourceGroup:
+
+# Modify how DNS records are sychronized between sources and providers (options: sync, upsert-only )
+policy: upsert-only
```
default value of ExternalDNS since `v0.3.0` is `--policy=sync`. I think we should use that as a default here as well.

sstarcher replied:
if the interest is using the default would it be best to comment out this line

rk295 replied:
My (slight) preference here is to explicitly set it, that prevents a situation where a user bumps the version of `external-dns` which may change it's default in the future meaning the state the chart leaves the service is would be unknown, i.e. a user would have to go and read the `external-dns` docs to discover the default for *their* chosen version.  I'm not hugely tided to that view though :)

rk295 replied:
D'oh, I've misspelt it, sink. It sync. It's late! Will fix up tomorrow.

linki commented on incubator/external-dns/templates/deployment.yaml:
```diff
@@ -0,0 +1,70 @@
+        - --registry={{ .Values.registry }}
+        - --txt-owner-id={{ .Values.txtOwnerId }}
+        {{- if .Values.compatibility }}
+        - --compatibility={{ .Values.compatibility }}
```
I think it is possible to not guard this. We could then uncomment `compatibility:` in values.yaml. This would lead to something like:

```yaml
args:
- --compatibility=
```

in the rendered manifest when no compatibility is used, which does work as expected (no compatibility). This approach works for other flags as well, e.g. `--namespace` (see https://github.com/kubernetes/charts/blob/master/stable/chaoskube/values.yaml#L23).

However, it clutters the manifest a bit but would cleanup the template and values.yaml. Not sure about this, let me know what you think @rk295.

rk295 replied:
My (again, slight) preference is to have the manifest as clean as possible, i.e. only those options which the user explicitly set.  Happy to do as you suggest though, if you think that is better.

linki replied:
Not sure either. It's your call.

linki commented on incubator/external-dns/templates/deployment.yaml:
```diff
@@ -0,0 +1,70 @@
+        {{- end }}
+        {{- if .Values.source.service }}
+        - --source=service
+        {{- end }}
```
iterating over `.Values.sources` and printing `--source={current iterator value}` seems nicer and more extensible to me.

sstarcher commented:
@rk295 sounds good on skipping the RBAC portion.  If you can add the ServiceAccount stuff it would be easy for a user to bind extra permissions to the pod.

It would look like
https://github.com/kubernetes/charts/blob/2f5852105ed27f1c2ffb701b2482a83834697d7b/incubator/istio/templates/ca-deployment.yaml#L25

and
https://github.com/kubernetes/charts/blob/2f5852105ed27f1c2ffb701b2482a83834697d7b/incubator/istio/templates/ca-service-account.yaml

sstarcher commented on incubator/external-dns/templates/deployment.yaml:
```diff
@@ -0,0 +1,70 @@
+    type: Recreate
+  template:
+    metadata:
+      labels: {{ include "labels" . | indent 8 }}
```
this will include the chart: label that will include the chart version.  This will make upgrades fail.

rk295 replied:
Is there a 'standard' set of labels charts should apply? It makes sense if they are uniform to me.

sstarcher replied:
generally 4 labels are applied to the deployment and only 3 to the pods.

sstarcher replied:
the biggest thing that matters is the label applied to the pod should not change on an upgrade and it will if you include the version.

rk295 replied:
👍

sstarcher commented on incubator/external-dns/templates/deployment.yaml:
```diff
@@ -0,0 +1,70 @@
+    metadata:
+      labels: {{ include "labels" . | indent 8 }}
+    spec:
+      containers:
```
should include the ability to apply resource restrictions. https://github.com/kubernetes/charts/blob/master/stable/uchiwa/templates/deployment.yaml#L21

rk295 committed:
Moving to stable

rk295 committed:
Some review fixes

rk295 commented:
@sstarcher & @linki - I've addressed a bunch of the comments above, a few still to sort though.

rk295 committed:
Typo

SamClinckspoor commented:
If you want to include RBAC, we use the following RBAC setup for external-dns. We don't have a lot of runtime with it yet but so far it seems to do the job:
```yaml
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: external-dns
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: external-dns
subjects:
  - kind: ServiceAccount
    name: external-dns
    namespace: default
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: external-dns
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRole
metadata:
  name: external-dns
rules:
- apiGroups:
  - ""
  - "extensions"
  resources:
  - services
  - ingresses
  verbs:
  - list
```

linki commented on stable/external-dns/README.md:
```diff
@@ -42,7 +42,7 @@ The following tables lists the configurable parameters of the consul chart and t
 | `domainFilter`         | Limit possible target zones by a domain suffix (optional).                                                    | `disabled`                     |
 | `azureResourceGroup`   | When using the Azure provider, override the Azure resource group to use (optional).                           | `disabled`                     |
-| `policy`               | Modify how DNS records are sychronized between sources and providers (options: sync, upsert-only ).           | `upsert-only`                  |
+| `policy`               | Modify how DNS records are sychronized between sources and providers (options: sync, upsert-only ).           | `sink`                  |
```
:satisfied:

linki commented on stable/external-dns/values.yaml:
```diff
@@ -0,0 +1,79 @@
+# azureResourceGroup:
+
+## Modify how DNS records are sychronized between sources and providers (options: sync, upsert-only )
+# policy: sink
```
reminder: change to `sync`

linki commented on stable/external-dns/values.yaml:
```diff
@@ -0,0 +1,79 @@
+registry: txt
+
+## When using the TXT registry, a name that identifies this instance of ExternalDNS (default: default)
+txtOwnerId: my-identifier
```
Since we're using helm could we have a smarter default value here? maybe the release name?

rk295 replied:
@linki yep, I like that plan. I'll do that.

rk295 replied:
QQ about that, how do I reference `{{ .Release.Name }}` in `values.yaml`? Or should I default it in the actual template?

linki replied:
Oh yes, I think you have to do it in the template and leave it blank here.

linki commented on stable/external-dns/values.yaml:
```diff
@@ -0,0 +1,79 @@
+## The registry implementation to use to keep track of DNS record ownership (default: txt, options: txt, noop )
+registry: txt
+
+## When using the TXT registry, a name that identifies this instance of ExternalDNS (default: default)
```
> (default: default)

seems odd

linki commented on stable/external-dns/values.yaml:
```diff
@@ -0,0 +1,79 @@
+# txtPrefix:
+
+## The interval between two consecutive synchronizations in duration format.
+# interval: 1
```
I think `1` wouldn't be valid. It has to be something like `1m`.

rk295 replied:
Correct, I've fixed.

linki commented on stable/external-dns/values.yaml:
```diff
@@ -0,0 +1,79 @@
+# metricsAddress: 127.0.0.1:7979
+
+## When enabled, increases the logging output for debugging purposes.
+debug: false
```
defining `debug: false` and at the same time commenting out `dryRun` doesn't seem right. Let's pick either approach.

rk295 replied:
Agreed, I've comment both on the basis they then won't end up in the manifest, but the user can set them if desired.

linki commented on stable/external-dns/values.yaml:
```diff
@@ -0,0 +1,79 @@
+# interval: 1
+
+## When enabled, exits the synchronization loop after the first iteration.
+# once:
```
Setting `once: true` is very unlikely to be used by a helm user. Maybe we can just leave it out.

rk295 replied:
Agreed, I just blindly added support for all the options without thinking that one through :)

rk295 committed:
More review comments

sstarcher commented on stable/external-dns/README.md:
```diff
@@ -0,0 +1,70 @@
+| `image.repository`     | Repository to pull the container from.                                                                        | `registry.opensource.zalan.do` |
+| `image.tag`            | Container image tag.                                                                                          | `v0.3.0`                       |
+| `master`               | The Kubernetes API server to connect to.                                                                      | `auto-detect`                  |
+| `source.fake`          | Boolean, should this resource be monitored.                                                                   | `false`                        |
```
This is actually a list in the values.yaml
```
source:
 - fake
 - ingress
```

jrnt30 commented on stable/external-dns/README.md:
````diff
@@ -0,0 +1,70 @@
+To install the chart with the release name `my-release`:
+
+```bash
+$ helm install --name my-release incubator/external-dns
````
Currently this PR is into `stable/`

rk295 replied:
👍

jrnt30 commented on stable/external-dns/templates/_helpers.tpl:
```diff
@@ -0,0 +1,23 @@
+We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
+*/}}
+{{- define "fullname" -}}
+{{- $name := default .Chart.Name .Values.nameOverride -}}
```
I don't think there has been great consolidation on this, however when you have template helper methods and you include a chart as a dependency, these names actually get added to the same global space and collide. In the absence of a good "common chart" I would suggest prefixing these helper methods with the chart name or something else similarly unique so they don't have the potential to collide with one another. 

Ex. 
```
{{- define "external-dns.fullname" -}}
{{- define "external-dns.labels" -}}
```

rk295 replied:
I missed this, now added.

ghost commented:
Any blockers left?

taylorsmithgg commented:
Just tested the PR, this is the only output:
```
time="2017-07-17T03:52:47Z" level=fatal msg="flag parsing error: enum value must be one of sync,upsert-only, got ''"
```

rk295 commented:
@taylorsmithgg my bad, since the PR work has had me on other things. I'll get back to - and tidy this up - in the next day or so.

linki commented on stable/external-dns/Chart.yaml:
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
+description: Deploys external-dns into a kubernetes cluster
```
Let's use a better description here. Maybe something like the one on the repo itself:

> Configure external DNS servers (AWS Route53, Google CloudDNS and others) for Kubernetes Ingresses and Services

rk295 replied:
👍

linki commented on stable/external-dns/Chart.yaml:
```diff
@@ -0,0 +1,9 @@
+description: Deploys external-dns into a kubernetes cluster
+name: external-dns
+version: 0.0.1
+appVersion: 0.3.0
```
Let's go with version `0.4`. As the final version isn't released yet you can use `v0.4.0-alpha.1` for that matter. I'll make sure it gets released before the chart is done.

We changed some flags in `v0.4` and then we don't have to change them again.

rk295 replied:
👍 I've gone for v0.4.0-alpha.2 if thats ok? I couldn't find any flags that had changed, do you have a link to any docs about that?

p.s. I'll push my changes shortly.

rk295 committed:
Reflecting the change to a list

rk295 commented:
@taylorsmithgg just looked into this and policy has to be either `upsert-only` or `sync`, I had it commented out in `values.yaml` so an empty `--policy=` option was being passed. I'm now defaulting to `upsert-only`, allowing the user of the chart to override if required.  I'll push that change shortly.

rk295 committed:
Minor changes to version and description.

rk295 committed:
This needs to be set to something.

rk295 committed:
reflecting the change in values.yaml from the previous commit.

rk295 committed:
Reflecting the move to `stable/`.

rk295 committed:
Preventing global namespace collisions.

rk295 committed:
Adding service account support.

rk295 commented:
@sstarcher I believe I've addressed the service account with the most recent commit, could you check and confirm please?

sstarcher commented:
@rk295 excellent, I'll review it tomorrow thanks for all of the hard work.

sstarcher commented on stable/external-dns/values.yaml:
```diff
@@ -0,0 +1,84 @@
+#    cpu: 10m
+
+# Should the chart create and use a service account. 
+# serviceAccount: false
```
I would recommend ```serviceAccountEnabled: false```

I originally assumed this was the name of the service account.

rk295 replied:
kk, just pushed a commit with that change.

sstarcher commented:
LGTM - lets get this in and add to it as needed

rk295 committed:
Renaming to avoid confusion

mgoodness commented on stable/external-dns/templates/deployment.yaml:
```diff
@@ -0,0 +1,70 @@
+        - --policy={{ .Values.policy }}
+        - --registry={{ .Values.registry }}
+        - --txt-owner-id={{default .Release.Name .Values.txtOwnerId }}
+        {{- if .Values.compatibility }}
```
Does each of the following arguments need to have its own value? I'm generally in favor of providing only the arguments that are commonly required to run the application, then providing an `extraArgs` map for users to provide additional ones. See [here](https://github.com/kubernetes/charts/blob/master/stable/prometheus/templates/server-deployment.yaml#L51) and [here](https://github.com/kubernetes/charts/blob/master/stable/kube2iam/templates/daemonset.yaml#L36) as examples.

My concern is that the list of options will continue growing, regardless of how commonly-used they are, making the chart unwieldy to new users. Better to keep it simple yet extendable.

rk295 replied:
Agreed, I prefer this approach. I've addressed all the other outstanding comments and will have a little tinker tonight to see if I can make this work and to pick a sane set of default options.

rk295 replied:
Oh! that was far easier than I thought. Thanks for the links to examples :) All comments addressed now.

mgoodness commented on stable/external-dns/templates/deployment.yaml:
```diff
@@ -0,0 +1,70 @@
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}"
+spec:
+  strategy:
+    type: Recreate
```
Is there an advantage to explicitly setting the strategy to `Recreate` rather than not providing one, thus defaulting to `RollingUpdate`?

rk295 replied:
I honestly don't know why I had this set to `Recreate`.  I'd prefer the default too, changed.

linki replied:
The reason we run it this way is because ExternalDNS shouldn't run concurrently. When usign the rolling updating this can happen though, whereas `Recreate` makes sure the current pod is fully terminated before the next one spawns (at least this worked fine on Kube 1.7).

However, the impact of a temporary concurrent run shouldn't have any harming effect, so 👍 .

rk295 replied:
Aha, @linki so that's where I got that from, good to know. I'm happy to flip it back if you think thats better? Your call.

mgoodness commented on stable/external-dns/templates/deployment.yaml:
```diff
@@ -0,0 +1,70 @@
+---
```
Delete. I've seen this cause problems.

rk295 replied:
👍

mgoodness commented on stable/external-dns/templates/deployment.yaml:
```diff
@@ -0,0 +1,70 @@
+      {{- end }}
+      containers:
+      - name: {{ .Chart.Name }}
+        image: "{{ .Values.image.repository }}/{{.Values.image.name}}:{{ .Values.image.tag }}"
```
Merge `image.name` into `image.repository`. We don't need both.

rk295 replied:
👍

mgoodness commented on stable/external-dns/README.md:
```diff
@@ -0,0 +1,70 @@
+
+## Configuration
+
+The following tables lists the configurable parameters of the consul chart and their default values.
```
`consul` -> `external-dns`

rk295 replied:
👍

mgoodness commented on stable/external-dns/values.yaml:
```diff
@@ -0,0 +1,84 @@
+##
+# podAnnotations: {}
+
+## CPU and Memory limit and request for Uchiwa
```
`Uchiwa` -> `external-dns`

rk295 replied:
👍

mgoodness reviewed (changes requested):
This looks pretty good to me. A few minor changes and a couple questions to address. Then we should be able to get it merged pretty quickly.

mgoodness commented:
Regarding RBAC, I'll happily add that (and tinker a bit with the ServiceAccount) once this is merged.

jrnt30 commented on stable/external-dns/templates/deployment.yaml:
```diff
@@ -0,0 +1,70 @@
+        {{- if .Values.compatibility }}
+        - --compatibility={{ .Values.compatibility }}
+        {{- end }}
+        {{- if .Values.debug }}
```
Debug is a bool flag, setting it to a value explicitly does not work

rk295 replied:
lol, good spot. Me being to hasty I expect!

jrnt30 commented on stable/external-dns/templates/deployment.yaml:
```diff
@@ -0,0 +1,70 @@
+        {{- if .Values.domainFilter }}
+        - --domain-filter={{ .Values.domainFilter }}
+        {{- end }}
+        {{- if .Values.dryRun }}
```
Dry run is also a bool flag, setting it to a value explicitly does not work

rk295 replied:
As above!

jrnt30 commented on stable/external-dns/values.yaml:
```diff
@@ -0,0 +1,84 @@
+
+# If creating a service account, you can optionally override the name used.
+# Defaults to .Release.Name-external-dns-sa. Note: Release.Name is always 
+# included to prevent collisions.
```
The way the helper works, this doesn't actually override the full service name it just overrides the suffix. That may be fine, but it was slightly confusing upon first use

rk295 replied:
I'm open to suggestions on this one. I pondered a few ways of doing it. I guess if there is some sort of (formal or otherwise) 'standard' used in Charts I'd look to follow that.

jordanjennings commented on stable/external-dns/Chart.yaml:
```diff
@@ -0,0 +1,9 @@
+description: Configure external DNS servers (AWS Route53, Google CloudDNS and others) for Kubernetes Ingresses and Services
+name: external-dns
+version: 0.0.1
+appVersion: v0.4.0-alpha.2
```
Release 0.4.1 is out now: https://github.com/kubernetes-incubator/external-dns/releases

rk295 replied:
👍

rk295 committed:
Removing the this because it can cause problems.

rk295 committed:
This was unintended, removing lets K8s use the default.

rk295 committed:
Review fixes.

rk295 committed:
Moving the majority of these into `extraArgs`.

I've picked the set of command line options we use in our rig, i.e. the
ones where we don't use the defaults. I /think/ this is a sane set.

linki commented on stable/external-dns/README.md:
```diff
@@ -0,0 +1,56 @@
+| `extaArg`              | Optional object of extra args, as `name`: `value` pairs. Where the name is the command line arg to external-dns. | ``                                                           |
+| `image.name`           | Container image name (Including repository name if not `hub.docker.com`).                                        | `registry.opensource.zalan.do/teapot/external-dns`           |
+| `image.pullPolicy`     | Container pull policy.                                                                                           | `IfNotPresent`                                               |
+| `image.tag`            | Container image tag.                                                                                             | `v0.3.0`                                                     |
```
This is actually `v0.4.2`.

rk295 replied:
👍

linki commented on stable/external-dns/README.md:
```diff
@@ -0,0 +1,56 @@
+| `image.tag`            | Container image tag.                                                                                             | `v0.3.0`                                                     |
+| `podAnnotations`       | Additional annotations to apply to the pod.                                                                      | `None`                                                       |
+| `policy`               | Modify how DNS records are sychronized between sources and providers (options: sync, upsert-only ).              | `upsert-only`                                                |
+| `provider`             | The DNS provider where the DNS records will be created (options: aws, google, inmemory, azure ).                 | `aws`                                                        |
```
Full list of supported providers as of `v0.4`: `aws, google, azure, cloudflare, digitalocean, inmemory`

rk295 replied:
👍

linki commented:
Just found two minor inconsistencies in the readme. Then LGTM 👍

rk295 committed:
Minor tweaks

rk295 commented:
@linki both those addressed now.

linki reviewed (approved):
lgtm

mgoodness committed:
Conform to standard RBAC pattern

mgoodness committed:
Update NOTES.txt

mgoodness committed:
Update README

mgoodness commented:
Made a few changes of my own, mostly just to bring the chart up to spec with our patterns. Biggest changes are that I renamed `source` to `sources` and `domainFilter` to `domainFilters` since both can be lists.

If you don't object to the changes, I'll get this merged right away and work on RBAC & IAM.

BTW, kudos for the `labels` helper function! Very nice way to DRY.

mgoodness commented:
/ok-to-test

mgoodness commented:
/lgtm

mgoodness committed:
Make domainFilters a list

rk295 commented:
@mgoodness 👍 to your changes. Good spot with `domainFilter`. I'm excited to get this in, I can finally remove the fork of it from our private repo.

rk295 commented:
Re your first comment, it does not come up without a key. I'll look at the chart you linked to and try and handle that more gracefully!
